src/Featurizers/TruncatedSVDFeaturizer.h

```diff
@@ -0,0 +1,382 @@
+    if ((this->_singularvalues - other._singularvalues).norm() > 0.000001f || (this->_singularvectors - other._singularvectors).norm() > 0.000001f)
+        return false;
+
+    return true;
+}
+
+template <typename MatrixT>
+void TruncatedSVDTransformer<MatrixT>::save(Archive &ar) const /*override*/ {
+    // Version
+    Traits<std::uint16_t>::serialize(ar, 1); // Major
+    Traits<std::uint16_t>::serialize(ar, 0); // Minor
+
+    // Data
+    Traits<decltype(_singularvalues)>::serialize(ar, _singularvalues);
+    Traits<decltype(_singularvectors)>::serialize(ar, _singularvectors);
+}
+
+// ----------------------------------------------------------------------
+// |
+// |  TruncatedSVDEstimator
+// |
+// ----------------------------------------------------------------------
+template <typename MatrixT, size_t MaxNumTrainingItemsV>
+TruncatedSVDEstimator<MatrixT, MaxNumTrainingItemsV>::TruncatedSVDEstimator(AnnotationMapsPtr pAllColumnAnnotations, size_t colIndex) :
+    BaseType("SVDEstimatorImpl", std::move(pAllColumnAnnotations)),
+    _colIndex(
+        std::move(
+            [this, &colIndex](void) -> size_t & {
+                if(colIndex >= this->get_column_annotations().size())
+                    throw std::invalid_argument("colIndex");
+
+                return colIndex;
+            }()
+        )
+    ) {
+}
+
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+template <typename MatrixT,size_t MaxNumTrainingItemsV>
+bool TruncatedSVDEstimator<MatrixT, MaxNumTrainingItemsV>::begin_training_impl(void) /*override*/ {
+    return true;
+}
+
+template <typename MatrixT, size_t MaxNumTrainingItemsV>
+void TruncatedSVDEstimator<MatrixT, MaxNumTrainingItemsV>::complete_training_impl(void) /*override*/ {
+    //the following code in this function is introduced from RedSVD
+    typedef typename MatrixT::Scalar Scalar;
+    typedef typename Eigen::Matrix<Scalar, Eigen::Dynamic, Eigen::Dynamic> DenseMatrix;
+
+    Eigen::Index rank = (_matrix.rows() < _matrix.cols()) ? _matrix.rows() : _matrix.cols();
+
+    // Gaussian Random Matrix for _matrix^T
+    DenseMatrix O(_matrix.rows(), rank);
+    sample_gaussian(O);
+
+    // Compute Sample Matrix of _matrix^T
+    DenseMatrix Y = _matrix.transpose() * O;
+
+    // Orthonormalize Y
+    gram_schmidt(Y);
+
+    // Range(B) = Range(_matrix^T)
+    DenseMatrix B = _matrix * Y;
+
+    // Gaussian Random Matrix
+    DenseMatrix P(B.cols(), rank);
+    sample_gaussian(P);
+
+    // Compute Sample Matrix of B
+    DenseMatrix Z = B * P;
+
+    // Orthonormalize Z
+    gram_schmidt(Z);
+
+    // Range(C) = Range(B)
+    DenseMatrix C = Z.transpose() * B;
+
+    Eigen::JacobiSVD<DenseMatrix> svdOfC(C, Eigen::ComputeThinV);
+
+    // C = USV^T
+    // A = Z * U * S * V^T * Y^T()
+    this->_singularValues = svdOfC.singularValues();
```

Comment:
It looks like _matrix isn't used anymore. We should free any memory associated with that object.